
## LAB REPORT 1 Description of Assignment:

For this lab report, you will be working with the titanic ml dadataset. You would use this dataset to learn to wrangle dataframes.


## Data Description/Challenges:

The sinking of the Titanic is one of the most infamous shipwrecks in history.
On April $15$, $1912$, during her maiden voyage, the widely considered “unsinkable” RMS Titanic sank after colliding with an iceberg. Unfortunately, there weren’t enough lifeboats for everyone onboard, resulting in the death of $1502$ out of $2224$ passengers and crew.

In this challenge, you are tasked to do some wrangling on the data and find ways to understand the data a bit more. This aassignment is due Thursday, January 30, 2025


#### 1. import the necessary libraries

In [ ]:
import pandas as pd

#### 1. import the necessary libraries

#### 2.  Load the train data in a variable name called "train"

In [ ]:
train = pd.read_csv("/content/test.csv")

#### 3.  Load the test data in a variable name called "test"

In [ ]:
test = pd.read_csv("/content/test.csv")

#### 4.  Check the shape of the train and test data

In [ ]:
print("Shape of the test data:", test.shape)

Shape of the test data: (418, 11)


In [ ]:
print("Shape of the train data:", train.shape)

Shape of the train data: (418, 11)


#### 5.  Comment on how many rows and columns the train and test data both have

In [ ]:
#Rows and Columns of the train data
print("Rows of the train data:", train.shape[0])

#Columns of the train data
print("Columns of the train data:", train.shape[1])


Rows of the train data: 418
Columns of the train data: 11


In [ ]:
#Rows and Columns of the test data
print("Rows of the test data:", test.shape[0])

#Columns of the test data
print("Columns of the test data:", test.shape[1])

Rows of the test data: 418
Columns of the test data: 11


The train data has 418 rows and 11 columns

The test data has 418 rows and 11 columns

#### 6.  Check the last 5 rows of the train and test data respectively.

In [ ]:
#Checking the last 5 rows of the train data
train.tail(5)

PassengerId  Pclass                          Name     Sex   Age  SibSp  \
413         1305       3            Spector, Mr. Woolf    male   NaN      0   
414         1306       1  Oliva y Ocana, Dona. Fermina  female  39.0      0   
415         1307       3  Saether, Mr. Simon Sivertsen    male  38.5      0   
416         1308       3           Ware, Mr. Frederick    male   NaN      0   
417         1309       3      Peter, Master. Michael J    male   NaN      1   

     Parch              Ticket      Fare Cabin Embarked  
413      0           A.5. 3236    8.0500   NaN        S  
414      0            PC 17758  108.9000  C105        C  
415      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416      0              359309    8.0500   NaN        S  
417      1                2668   22.3583   NaN        C

In [ ]:
#Checking the last 5 rows of the test data
test.tail(5)

PassengerId  Pclass                          Name     Sex   Age  SibSp  \
413         1305       3            Spector, Mr. Woolf    male   NaN      0   
414         1306       1  Oliva y Ocana, Dona. Fermina  female  39.0      0   
415         1307       3  Saether, Mr. Simon Sivertsen    male  38.5      0   
416         1308       3           Ware, Mr. Frederick    male   NaN      0   
417         1309       3      Peter, Master. Michael J    male   NaN      1   

     Parch              Ticket      Fare Cabin Embarked  
413      0           A.5. 3236    8.0500   NaN        S  
414      0            PC 17758  108.9000  C105        C  
415      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416      0              359309    8.0500   NaN        S  
417      1                2668   22.3583   NaN        C

#### 7.  Do a describitive of both the train and test data. Comment on the results from both data.

Both the last 5 rows of the train and test data have the same values for the PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked columns.

#### 8.  What are the column names in the train and test data respectively. Do you have any column in train which is absent in the test

The column names in the train data are PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin and Embarked.

The column names in the test data are PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked.

No, there's no column in train which is absent in test.

#### 9.  Counts for the passengers who survived. Do this for only the train data

In [ ]:
#Counts for the passengers who survived in the train data
train_count = train['Name'].count()
print(train_count)

418


#### 10.  Extract only the titles from the name columns in the train
**Hint**: Check on using regular expression to extract the title from the name column.

In [ ]:
#Extracting only the titles from the name columns in the train data
titles = ['Mlle.', 'Major.', 'Master.' ,'Col.', 'Sir.', 'Don.', 'Mme.', 'Jonkheer.', 'Lady.', 'Capt.', 'Countess.', 'Ms.', 'Dona.']
train_titles_extracted = train[train['Name'].str.contains('|'.join(titles), case=False, na=False)]
print(train_titles_extracted)

     PassengerId  Pclass                                               Name  \
21           913       3                          Olsen, Master. Artur Karl   
23           915       1                    Williams, Mr. Richard Norris II   
43           935       2              Corbett, Mrs. Walter H (Irene Colvin)   
55           947       3                               Rice, Master. Albert   
64           956       1                        Ryerson, Master. John Borie   
77           969       1  Cornell, Mrs. Robert Clifford (Malvina Helen L...   
80           972       3                               Boulos, Master. Akar   
88           980       3                            O'Donoghue, Ms. Bridget   
89           981       2                        Wells, Master. Ralph Lester   
118         1010       1                               Beattie, Mr. Thomson   
131         1023       1                          Gracie, Col. Archibald IV   
151         1043       3                            

**Note:** You can only do this question after doing the previous question
 #### 11. Create a new column and call it "Title" to save the extracted titles above

In [ ]:
#Saving the titles above into the "Title" column
# Regular expression to extract titles from the Name column
train['Title'] = train['Name'].str.extract(r'(\b(?:Mlle.|Major.|Master.|Col.|Sir.|Don.|Mme.|Jonkheer.|Lady.|Capt.|Countess.|Ms.|Dona.)\b)', expand=True)
#'Mlle', 'Major', 'Master' ,'Col', 'Sir', 'Don', 'Mme', 'Jonkheer', 'Lady', 'Capt', 'Countess', 'Ms', 'Dona'

# Displaying the first few rows to verify the new Title column
print(train[['Name', 'Title']].head(20))


                                                 Name Title
0                                    Kelly, Mr. James   NaN
1                    Wilkes, Mrs. James (Ellen Needs)   NaN
2                           Myles, Mr. Thomas Francis   NaN
3                                    Wirz, Mr. Albert   NaN
4        Hirvonen, Mrs. Alexander (Helga E Lindqvist)   NaN
5                          Svensson, Mr. Johan Cervin   NaN
6                                Connolly, Miss. Kate   NaN
7                        Caldwell, Mr. Albert Francis   NaN
8           Abrahim, Mrs. Joseph (Sophie Halaut Easu)   NaN
9                             Davies, Mr. John Samuel   NaN
10                                   Ilieff, Mr. Ylio   NaN
11                         Jones, Mr. Charles Cresson   NaN
12      Snyder, Mrs. John Pillsbury (Nelle Stevenson)   NaN
13                               Howard, Mr. Benjamin   NaN
14  Chaffee, Mrs. Herbert Fuller (Carrie Constance...   NaN
15      del Carlo, Mrs. Sebastiano (Arge

#### 12.  Use this dictionary to change the title names in the train.Title created column

mapping = {'Mlle': 'Miss', 'Major': 'Mr', 'Col': 'Mr', 'Sir': 'Mr',
           'Don': 'Mr', 'Mme': 'Mrs', 'Jonkheer': 'Mr', 'Lady': 'Mrs',
           'Capt': 'Mr', 'Countess': 'Mrs', 'Ms': 'Miss', 'Dona': 'Mrs'}

In [ ]:
#Changing the title names in the train in the title created column
mapping = {'Mlle': 'Miss', 'Major': 'Mr', 'Col': 'Mr', 'Sir': 'Mr', 'Don': 'Mr', 'Mme': 'Mrs', 'Jonkheer': 'Mr', 'Lady': 'Mrs', 'Capt': 'Mr', 'Countess': 'Mrs', 'Ms': 'Miss', 'Dona': 'Mrs'}

train['Title'] = train['Title'].replace(mapping)
print(train[['Name', 'Title']].head())

                                           Name Title
0                              Kelly, Mr. James   NaN
1              Wilkes, Mrs. James (Ellen Needs)   NaN
2                     Myles, Mr. Thomas Francis   NaN
3                              Wirz, Mr. Albert   NaN
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   NaN


#### 13.  Value counts after changing some titles

In [ ]:
#The new value counts after changing some titles
value_count = train['Title'].value_counts()
print(value_count)

Title
Mrs    1
Name: count, dtype: int64


#### 14. Group data by class and view the averages for each column

In [ ]:
class_averages = train.groupby('Pclass').mean(numeric_only=True)
print(class_averages)

        PassengerId        Age     SibSp     Parch       Fare
Pclass                                                       
1       1098.224299  40.918367  0.476636  0.383178  94.280297
2       1117.935484  28.777500  0.376344  0.344086  22.202104
3       1094.178899  24.027945  0.463303  0.417431  12.459678


#### 15.  What do you observe from the  above groupby table? Discuss and write down your thoughts

I observed that the Titanic dataset contains three classes:
1st class: Wealthier passengers.
2nd class: Middle-class passengers.
3rd class: Lower-class passengers or those with the cheapest tickets.

Under normal circumstances, the average of each column should have been a single value but there are three average values in each table which means that the titanic dataset contains three classes as described above.


#### 16.  Group by Pclass and sex and find the mean


In [ ]:
class_and_sex_averages = train.groupby(['Pclass', 'Sex']).mean(numeric_only=True)
print(class_and_sex_averages)

               PassengerId        Age     SibSp     Parch        Fare
Pclass Sex                                                           
1      female  1104.080000  41.333333  0.560000  0.500000  115.591168
       male    1093.087719  40.520000  0.403509  0.280702   75.586551
2      female  1111.200000  24.376552  0.533333  0.766667   26.438750
       male    1121.142857  30.940678  0.301587  0.142857   20.184654
3      female  1085.722222  23.073400  0.583333  0.597222   13.735129
       male    1098.349315  24.525104  0.404110  0.328767   11.826350


#### 17. **Missing Values**

#### 18.  How many missing values are in train data  and test respectively.


In [ ]:
#Missing values in the train data
train_missing_values = train.isnull().sum()
print(train_missing_values)

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
Title          417
dtype: int64


In [ ]:
#Missing values in the test data
test_missing_values = test.isnull().sum()
print(test_missing_values)

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64


#### 19.  Which of columns in the train data had missing values?

The columns in the train data that had missing values are Age, Fare, Cabin, Title.

#### 20.  Which of columns in the test data had missing values?


The columns in the test data that had missing values are Age, Fare, Cabin.

#### 21.  First drop all missing values and check the counts for missing value again

**Remember** Do not run the code on dropping all missing values the second time

#### 22.  Reload the data and drop the missing values in age with the mean


#### 23. Drop the missing values with the Embarked with the mode


#### 24. Merge the two dataframes and called it df.Check to the number of rows and columns and number of missing values

In [ ]:
#import pandas as pd
#df = pd.read_csv("file path of the csv file")
#